sys/net/application_layer/ng_zep/ng_zep.c

```diff
@@ -0,0 +1,1072 @@
+#include "vtimer.h"
+
+#include "net/ng_zep.h"
+
+#define ENABLE_DEBUG    (0)
+#include "debug.h"
+
+#define _EVENT_RX_STARTED       (1)
+#define _EVENT_RX_COMPLETE      (2)
+#define _RX_BUF_SIZE            (16U * sizeof(ng_pktsnip_t *))
+#define IEEE_802154_FCS_POLY    (0x8408)  /* x^16 + x^12 + x^5 + 1 for LSB first */
+
+static kernel_pid_t _pid = KERNEL_PID_UNDEF;
+static char _rx_stack[NG_ZEP_STACK_SIZE];
+static char _rx_buf_array[_RX_BUF_SIZE];
+static ringbuffer_t _rx_buf = RINGBUFFER_INIT(_rx_buf_array);
+
+/* ng_netdev driver definitions */
+static int _send(ng_netdev_t *dev, ng_pktsnip_t *pkt);
+static int _add_cb(ng_netdev_t *dev, ng_netdev_event_cb_t cb);
+static int _rem_cb(ng_netdev_t *dev, ng_netdev_event_cb_t cb);
+static int _get(ng_netdev_t *dev, ng_netconf_opt_t opt, void *value,
+                size_t max_len);
+static int _set(ng_netdev_t *dev, ng_netconf_opt_t opt, void *value,
+                size_t value_len);
+static void _isr_event(ng_netdev_t *dev, uint32_t event_type);
+
+static const ng_netdev_driver_t _zep_driver = {
+    _send,
+    _add_cb,
+    _rem_cb,
+    _get,
+    _set,
+    _isr_event
+};
+
+/* Function for the ZEP thread */
+void *_event_loop(void *args);
+
+/* Builds (uninitialized) ZEP packet according to configured version */
+static ng_pktsnip_t *_zep_hdr_build(ng_zep_t *dev, size_t size, bool ack);
+
+/* Fills ZEP header according to ng_zep_t configuration */
+static size_t _zep_hdr_fill(ng_zep_t *dev, ng_zep_hdr_t *hdr,
+                            size_t payload_len);
+
+/* Event handlers for ISR events */
+static void _rx_started_event(ng_zep_t *dev);
+
+/* IEEE 802.15.4 helper functions: TODO: generalize add to ng_ieee802154 */
+static size_t _make_data_frame_hdr(ng_zep_t *dev, uint8_t *buf,
+                                   ng_netif_hdr_t *hdr);
+static size_t _get_frame_hdr_len(uint8_t *mhr);
+ng_pktsnip_t *_make_netif_hdr(uint8_t *mhr);
+static uint16_t _calc_fcs(uint16_t fcs, const uint8_t *frame, uint8_t frame_len);
+
+kernel_pid_t ng_zep_init(ng_zep_t *dev, uint16_t src_port, ng_ipv6_addr_t *dst,
+                         uint16_t dst_port)
+{
+#if CPUID_ID_LEN
+    uint8_t cpuid[CPUID_ID_LEN];
+    uint32_t hash1, hash2;
+#endif
+
+    if (_pid != KERNEL_PID_UNDEF) {
+        DEBUG("zep: ZEP thread already running at pid=%" PRIkernel_pid "\n", _pid);
+        return -EEXIST;
+    }
+
+    if (dev == NULL) {
+        DEBUG("zep: dev was NULL\n");
+        return -ENODEV;
+    }
+
+    if ((dst == NULL) || (ng_ipv6_addr_is_unspecified(dst))) {
+        DEBUG("zep: dst (%s) was NULL or unspecified\n", dst);
+        return -ENOTSUP;
+    }
+
+    if (ng_netreg_lookup(NG_NETTYPE_UDP, src_port)) {
+        DEBUG("zep: port (%" PRIu16 ") already registered\n", src_port);
+        return -EADDRINUSE;
+    }
+
+    dev->driver = (ng_netdev_driver_t *)&_zep_driver;
+    dev->chan = NG_ZEP_DEFAULT_CHANNEL;
+    dev->pan = byteorder_btols(byteorder_htons(NG_ZEP_DEFAULT_PANID));
```

Comment:
A constant that a user is able to change. Think like this:

* if we are on a big-endian system `byteorder_htons()` has no effect and `byteorder_btols()` swaps it to little-endian
* if we are on a little-endian system `byteorder_htons()` swaps it to big-endian and then `byteorder_btols()` back to little-endian.

Since it *is* a constant the compiler normally optimizes this out, so I thought I'm on the safer side here.